keps/sig-node/3066-subsecond-probes/README.md

```diff
@@ -0,0 +1,859 @@
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md)
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+The Probe struct contains fields that specify seconds for intervals.
+Some users would like to have intervals less than or slightly greater
+than one second.
+
+## Motivation
+
+Better performance in the form of faster startups, more timely readiness checks...
+
+#### Knative
+
+For example, Knative will create Pods (via Deployment) and wait for them to become `Ready`
+when handling an HTTP request from an end-user. In this case, Pod readiness
+latency has a direct impact on HTTP request latency. (In the steady state,
+Knative will re-use an existing Pod, but this situation can happen on scale-up,
+and is guaranteed to happen on scale-from-zero.)
+
+### Goals
+
+An ability to specify probe durations that have more granularity, more specificity, than just zero,
+one, or n-seconds. Focus is on possibly faster readiness probes, for example every .5 seconds until succss
+and more granular use beyond existing defaults, for example 1.5 second duration instead of picking
+between 1 or 2 second duration.
+Add additional tests cases to the timeout test cases.
+Not breaking backwards compatibility with the V1 API.
+
+
+### Non-Goals
+
+V2 API for existing objects.
+Converting fields from `int32` to `resource.Quantity`.
+
+
+## Proposal
+
+TLDR:
+Add two new optional fields (of type `*int32`) to the Probe struct, which would be used to offset the second based time values in the Probe struct:
+
+- `PeriodMilliseconds`       // How often (in milliseconds) to offset PeriodSeconds when performing the probe (*** to reduce un-necessary resource usage, when periodMilliseconds is used to reduce the period to less than a second the offset is only used until success is reached then no longer).
```

Comment:
initialDelaySeconds 0 initialDelayMilliseconds  500
periodSeconds 5

first probe should be at 500mills .. then wait 5sec for second probe

Reply:
I thought this was describing a period of rapid polling until the probe succeeds? Initial delay is something different?

Reply:
Not sure we are in sync yet.. maybe a call?

The idea here is the initial delay and period, being defined as whole seconds, is too coarse.  Without creating a v2 api for probes and to keep probes backwards compatible.. this proposal is to introduce fine grained offsets for the these course fields. E.g. where it it known the initial delay needs to be at least 1.5 seconds the admin/developer must choose 2seconds today.. and with this feature may define 2seconds minus 500mills resulting in 1.5seconds with the gate on and 2seconds with the gate off..

Reply:
WRT rapid polling before success.. a limit must be set on the period to avoid the probe DOSing the node. This would apply to the budget discussion, as opposed to the overtly simple measure of limiting arguably expensive exec probes to up to twice as fast as they are today until success.. and cheaper probe types to 200mils (or up to 5x faster.) Similarly noting that it's not necessarily "faster" the author may simply prefer a value between two seconds instead of a first whole second or the first +1second.